File: runtime/CSlib/weakptr.cs
using System;

namespace bigloo
{
  public class weakptr: obj
  {
    public WeakReference mref;
    
    public weakptr( Object  data)
    {
      this.mref = new WeakReference(data);
    }

    public Object getData(){
      Object d = mref.Target;
      return d == null ? unspecified._unspecified : d;
    }

    public void setData(Object o){
      mref = new WeakReference(o);
    }

    public override void write( output_port p ) {
      p.write("#<weakptr:");
      foreign.write_object(getData(), p);
      p.write(">");
    }
  }
}

File: runtime/Jlib/weakptr.java
package bigloo;

import java.lang.ref.WeakReference;


public class weakptr extends obj {
  public WeakReference data;
  public WeakReference ref;

  public weakptr(final Object d, final Object r){
     data = new WeakReference(d);
     ref = new WeakReference(r);
  }

  public Object getData(){
    Object d = ref.get();
    return d == null ? unspecified.unspecified : d;
  }

  public void setData(Object o){
    data = new WeakReference(o);
  }

  public Object getRef(){
    Object d = data.get();
    return d == null ? unspecified.unspecified : d;
  }

  public void setRef(Object o){
    ref = new WeakReference(o);
  }

  public void write( final output_port p ) {
    p.write("#<weakptr:");
    foreign.write_object(getData(), p);
    p.write(">");
  }
}
